### ISSUE TITLE
Train without pretrained weights?

### ISSUE THREAD

**Issue by hhhyyeee** [MarcBotet/hamlet#6]
Hi, I'd like to ask you guys if it's possible to train HAMLET without pretrained weights.

I'm currently working with exp -1 option, therefore controlling everything with `config.py` file.
And since I wanted to train HAMLET without pretrained weights (`pretrained/mitb1_uda.pth`), I tried these 3 ways:

1) `pretrained_segmentator = None`
2) `pretrained_segmentator = ""`
3) `# pretrained_segmentator`

However it seems that in several functions and modules including `build_train_model` and `ModularUDADecoder` happen the KeyErrors, indicating that none of these 3 ways are working.

Can you give me any advice?

(+) Additionally, is this model weights file ([link](https://drive.google.com/drive/folders/1p3VoGrsQsI7wMCYeAb7nYEImW0oIrmMR)) pretrained in the source (CityScapes train) dataset? or is it just initialized or slightly warmed up? I couldn't find any detailed information about the weights.

Thank you.

**Comment by MarcBotet** [comment]
Hi, thank you for the interest in our project.

For the Hamlet scenario we always train assuming the model has been pretrained on the source dataset, so there is no support in not setting a pretrained weights.
However, I would recommend using SegFormer weights where the encoder is pretrained on ImageNet-1K. This can be found in the SegFormer github and can be used in the same way as mitb1_uda.pth. We found that using the pretrained SegFormer on ImageNet-1K instead of random weights is crucial to get competitive performance. 
In the case you truly need random initialization, let me know and I will try to help you.

The provided model weights are pretrained on CityScapes dataset using SegFormer pretrained weights for 100’000 iterations (selecting the checkpoint with the best validation accuracy) using a learning rate of 6 × 10−5, warm-up and linear decay scheduling. You can find further details about the implementation in the supp. material, in the same arxiv paper.

Let me know if you have further questions

**Comment by hhhyyeee** [comment]
Thank you so much for your answer. I completely understood!

I have one further question though.

When I tried to train HAMLET with default training settings according to `config.py`, the following datasets were fetched and used in training: `Cityscapes train [clean, 25mm, ..., 100mm, 200mm, 100mm, ..., 25mm, clean]`
(I guess the `Cityscapes clean` dataset is the one that has been used to warm up the model weights.)

I presumed that 25mm~200mm weather_datasets are used to train the model in online fashion, backpropagating in self-supervised way in CTTA purpose.
However, it seems a bit unfamiliar to me to use `Cityscapes clean` dataset every single epoch, because of these 2 reasons:
(1) The model has already been trained (warmed up) on the dataset.
(2) In CTTA setting, preventing catastrophic forgetting about the source dataset is one of the most important objectives, hence revisiting the source dataset over and over does not make much sense.

I'm sure there must be some mistakes I made understanding the whole training procedure.
Can you help me understand more clearly?

Thank you.

**Comment by MarcBotet** [comment]
Hi, we follow the online domain adaptation setup of [onda](https://arxiv.org/abs/2207.10667) where we also make use of a replay buffer with a subset of the source dataset.

**Comment by hhhyyeee** [comment]
Thank you so much for your kind reply.

**Comment by hhhyyeee** [comment]
Thank you so much for your kind reply.
